### PR TITLE
extract line reader handling decoding from harvester

### PR DESCRIPTION
**Pull request by urso**
split line reader from harvester. reader separates lines, applies the configured codec + while reading lines it keeps track of total bytes consumed from raw input stream (speak: correct file offset in bytes).

### CODE REVIEW

File: harvester/encoding.go [hunk 1]
```diff
@@ -8,41 +8,41 @@ import (
...
-	// utf8 (validate input) - shadow htmlindex utf8 codecs not validating input
-	"unicode-1-1-utf-8": trans(encoding.UTF8Validator),
-	"utf-8":             trans(encoding.UTF8Validator),
-	"utf8":              trans(encoding.UTF8Validator),
+	/*
+		// utf8 (validate input) - shadow htmlindex utf8 codecs not validating input
+		"unicode-1-1-utf-8": trans(encoding.UTF8Validator),
+		"utf-8":             trans(encoding.UTF8Validator),
+		"utf8":              trans(encoding.UTF8Validator),
+	*/
```
Comment (urso) [inline harvester/encoding.go]: TODO: create full encoding based on utf8 validator.

File: harvester/encoding.go [hunk 2]
```diff
@@ -8,41 +8,41 @@ import (
...
 
 	// utf16
-	"utf-16be-bom": enc(unicode.UTF16(unicode.BigEndian, unicode.UseBOM)),
+	// "utf-16be-bom": unicode.UTF16(unicode.BigEndian, unicode.UseBOM),
```
Comment (urso) [inline harvester/encoding.go]: decided to disable "utf-16be-bom" in this PR, as handling this codec needs some more logic on start. To be done in another PR.

File: harvester/log.go [hunk 3]
```diff
@@ -292,55 +294,37 @@ func lineEndingChars(line []byte) int {
...
-			str := new(string)
-			// Remove line endings
-			*str = buffer.String()[:bufferSize-lineEndingChars(segment)]
-			// Reset the buffer for the next line
-			buffer.Reset()
-			return str, bufferSize, nil
+		// test for no file updates longer than partialLineWaiting
+		if time.Since(*lastReadTime) >= partialLineWaiting {
+			// return partial line:
+			// XXX: we rather want to drop partial line (like orignial implementation?)
```
Comment (urso) [inline harvester/log.go]: opinions?

File: harvester/log.go [hunk 4]
```diff
@@ -65,10 +63,16 @@ func (h *Harvester) Harvest() {
 	// Load last offset from registrar
 	h.initOffset()
 
-	in := h.encoding(h.file)
+	// TODO: newLineReader uses additional buffering to deal with encoding and testing
+	//       for new lines in input stream. Simple 8-bit based encodings, or plain
+	//       don't require 'complicated' logic.
```
Comment (urso) [inline harvester/log.go]: TODO: but can be done in another PR. Current solution maybe not optimal for performance + memory usage, but not incorrect.

**Comment by urso** [comment]
requires elastic/libbeat#290

File: harvester/reader_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			continue
+		}
+		for i := range test.lines {
+			expected := test.lines[i]
+			actual := readLines[i]
+			assert.Equal(t, expected, actual)
+			assert.Equal(t, expectedCount[i], byteCounts[i])
+		}
+	}
+}
```
Comment (urso) [inline harvester/reader_test.go]: TODO: add tests for reading partial lines (missing EOL)

File: harvester/log.go (hunk 3 above)
Comment (ruflin) [inline harvester/log.go]: I would actually return partial lines. I think it is a rare case and it is better to have a line twice then not at all.

File: harvester/log.go [hunk 6]
```diff
@@ -292,55 +294,37 @@ func lineEndingChars(line []byte) int {
...
-func readLine(reader *bufio.Reader, buffer *bytes.Buffer, partialLineWaiting time.Duration) (*string, int, error) {
-
-	lastSegementTime := time.Now()
-	isPartialLine := true
-
+func readLine(
+	reader *lineReader,
+	lastReadTime *time.Time,
+	partialLineWaiting time.Duration,
+) (string, int, error) {
```
Comment (ruflin) [inline harvester/log.go]: we had *string before as it made it possible to differentiate between an empty line ("") and no line (nil). At the same time error should be set in case it is not a line.

File: harvester/reader.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/libbeat/common/streambuf"
+)
+
+// timedReader keeps track of last time bytes have been read from underlying
+// reader.
+type timedReader struct {
```
Comment (ruflin) [inline harvester/reader.go]: Instead of timedReader I would probably call it lineReader or something similar. My first impression with the name was that it is a reader that reads and a predefined timing.

File: harvester/reader.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+	nl        []byte
+	inBuffer  *streambuf.Buffer
+	outBuffer *streambuf.Buffer
+	inOffset  int // input buffer read offset
+	byteCount int // number of bytes decoded from input buffer into output buffer
+	decoder   transform.Transformer
+}
+
+const maxConsecutiveEmptyReads = 100
```
Comment (ruflin) [inline harvester/reader.go]: Is this value based on some assumptions you made? Or related to the BufferSize?

File: harvester/reader_test.go (hunk 5 above)
Comment (ruflin) [inline harvester/reader_test.go]: @urso There are some system tests for this already.

**Comment by ruflin** [comment]
I like this approach a lot as it simplifies the harvester, as in log.go it doesn't have to be dealt with buffering and encoding. I'm thinking if we should move reader.go in the long term to libbeat or even contribute parts to the Go source code.

File: harvester/log.go (hunk 3 above)
Comment (urso) [inline harvester/log.go]: before this PR, partial lines have been dropped, but input stream has been consumed.

This PR returns the partial line, but bytes are also consumes the bytes returned. That is the substring returned here will not be repeated in subsequent calls. I was not sure about wanted semantics here. Maybe we change this to repeat the complete line if any bytes have been read in meantime.

File: harvester/reader_test.go (hunk 5 above)
Comment (urso) [inline harvester/reader_test.go]: cool. As this encoding stuff is quite fragile we should check the combined coverage report and see if we're missing something.

File: harvester/log.go (hunk 6 above)
Comment (urso) [inline harvester/log.go]: the harvester worker is not using the returned line if error!=nil. And readLine reads until EOF or some bytes have been consumed.

File: harvester/reader.go (hunk 7 above)
Comment (urso) [inline harvester/reader.go]: Don't like timedReader either, but lineReader is no good name too.

this reader sits between the file input and the line reader. It's only purpose is to update an internal timestamp whenever any bytes have been read from file. The timestamp is used in readLine with partialLineWaiting to have some kind of 'keep-alive' ;)

File: harvester/reader.go (hunk 8 above)
Comment (urso) [inline harvester/reader.go]: I did follow source code of most functions used before PR in order to be as close in behavior to original implementation as possible. This variable + loop on read is 'inspired' by standard library.

**Comment by urso** [comment]
Not sure about moving to libbeat as is. Behavior is tailored for filebeat harvester (but quite generic, true). When moving we should improve on the interfaces in reader.go (naming is really bad).

File: harvester/log.go (hunk 3 above)
Comment (ruflin) [inline harvester/log.go]: If we drop the partial line, we definitively should retry later at the beginning of the line. What is easier to implement?

File: harvester/reader.go (hunk 7 above)
Comment (ruflin) [inline harvester/reader.go]: Not "urgent" at the moment. We can also rename it later when we have a better name.

File: harvester/reader.go (hunk 8 above)
Comment (ruflin) [inline harvester/reader.go]: :+1:

File: harvester/log.go (hunk 3 above)
Comment (urso) [inline harvester/log.go]: easiest is as is. When restarting at beginning of line we have to check line was updated since last partial return, else we might repeat lines quite a load of times.

Comment (ruflin) [inline harvester/log.go]: Ok, lets keep it for the moment like this.

File: harvester/log.go (hunk 4 above)
Comment (ruflin) [inline harvester/log.go]: I think this would mainly matter for really large lines.

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,10 +6,13 @@ All notable changes to this project will be documented in this file based on the
 ## [Unreleased](https://github.com/elastic/libbeat/compare/1.0.0-rc1...HEAD)
 
 ### Backward Compatibility Breaks
+- Removed utf-16be-bom encoding support
```
Comment (andrewkroh) [inline CHANGELOG.md]: From a user perspective, it would be nice to answer the "why" question or link to an issue.

Comment (urso) [inline CHANGELOG.md]: I will re-add utf-16-bom soonish.

**Comment by andrewkroh** [comment]
Looks good.
